filebeat/input/tcp/client.go

```diff
@@ -0,0 +1,118 @@
+package tcp
+
+import (
+	"bufio"
+	"net"
+	"strings"
+	"time"
+
+	"github.com/pkg/errors"
+
+	"github.com/elastic/beats/filebeat/harvester"
+	"github.com/elastic/beats/filebeat/util"
+	"github.com/elastic/beats/libbeat/beat"
+	"github.com/elastic/beats/libbeat/common"
+	"github.com/elastic/beats/libbeat/logp"
+)
+
+// Client is a remote client.
+type Client struct {
+	conn          net.Conn
+	forwarder     *harvester.Forwarder
+	done          chan struct{}
+	metadata      common.MapStr
+	splitFunc     bufio.SplitFunc
+	maxReadBuffer int64
+	timeout       time.Duration
+}
+
+// NewClient returns a new client instance for the remote connection.
+func NewClient(
+	conn net.Conn,
+	forwarder *harvester.Forwarder,
+	splitFunc bufio.SplitFunc,
+	maxReadBuffer int64,
+	timeout time.Duration,
+) *Client {
+	client := &Client{
+		conn:          conn,
+		forwarder:     forwarder,
+		done:          make(chan struct{}),
+		splitFunc:     splitFunc,
+		maxReadBuffer: maxReadBuffer,
+		timeout:       timeout,
+	}
+	client.cacheMetadata()
+	return client
+}
+
+// Handle is reading message from the specified TCP socket.
+func (c *Client) Handle() error {
+	defer c.conn.Close()
```

Comment:
I wonder if this `defer` should really be here or outside where the connection is opened or `Handle` is called.

Reply:
Make sense, I will that change when you are done with the review.